File: 2020/Swift/04-PassportProcessing/PassportProcessing.swift
#!/usr/bin/swift

// Passport Processing
// https://adventofcode.com/2020/day/4

import Foundation

func readFile(named name: String) -> [String] {
    let currentDirectoryURL = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
    let fileURL = URL(fileURLWithPath: name + ".txt", relativeTo: currentDirectoryURL)
    guard let content = try? String(contentsOf: fileURL, encoding: String.Encoding.utf8) else {
        print("Unable to read input file \(name)")
        print("Current directory: \(currentDirectoryURL)")
        return []
    }
    return content.components(separatedBy: .newlines)
}

func parsePassports(lines: [String]) -> [[String : String]] {
    var passports = [[String : String]]()
    var passport = [String : String]()
    for line in lines {
        guard !line.isEmpty else {
            passports.append(passport)
            passport = [String : String]()
            continue
        }
        let fields = parseFields(line: line)
        passport.merge(fields) { (_, new) in new }
    }
    return passports
}

func parseFields(line: String) -> [String : String] {
    var fields = [String : String]()
    let fieldsString = line.split(separator: " ")
    for field in fieldsString {
        let values = field.split(separator: ":")
        guard values.count == 2 else {
            continue
        }
        let key = String(values[0])
        let value = String(values[1])
        fields[key] = value
    }
    return fields
}

func hasRequiredFields(passport: [String : String]) -> Bool {
    let requiredFieldKeys = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
    let requiredFields = requiredFieldKeys.compactMap { passport[$0] }
    return requiredFields.count == requiredFieldKeys.count
}

func hasValidRequiredFields(passport: [String : String]) -> Bool {
    let requiredFields = [
        ("byr", { value in isValidYear(value, min: 1920, max: 2002) }),
        ("iyr", { value in isValidYear(value, min: 2010, max: 2020) }),
        ("eyr", { value in isValidYear(value, min: 2020, max: 2030) }),
        ("hgt", isValidHeight),
        ("hcl", isValidHairColor),
        ("ecl", isValidEyeColor),
        ("pid", isValidPassportId)]
    let hasValidFields = requiredFields.reduce(true) { (result, keyValidator) in
        let (key, isValid) = keyValidator
        return result && isValid(passport[key])
    }
    return hasValidFields
}

func isValidYear(_ string: String?, min: Int, max: Int) -> Bool {
    guard let string = string else { return false }
    guard let year = Int(string) else { return false }
    return min...max ~= year
}

func isValidHeight(_ string: String?) -> Bool {
    guard let string = string else { return false }
    guard let value = Int(string.dropLast(2)) else { return false }
    let validInches = 59...76
    let validCentimeters = 150...193
    return
        (string.hasSuffix("in") && validInches ~= value) ||
        (string.hasSuffix("cm") && validCentimeters ~= value)
}

func isValidHairColor(_ string: String?) -> Bool {
    guard let string = string else { return false }
    guard string.hasPrefix("#") else { return false }
    let value = string.dropFirst()
    let valid = CharacterSet(charactersIn: "0123456789abcdef")
    return value.unicodeScalars.allSatisfy { valid.contains($0) }
}

func isValidEyeColor(_ string: String?) -> Bool {
    guard let string = string else { return false }
    let valid: Set = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
    return valid.contains(string)
}

func isValidPassportId(_ string: String?) -> Bool {
    guard let string = string else { return false }
    guard string.count == 9 else { return false }
    let valid = CharacterSet.decimalDigits
    return string.unicodeScalars.allSatisfy { valid.contains($0) }
}

let passportStrings = readFile(named: "04-input")
let passports = parsePassports(lines: passportStrings)
let passportsWithRequiredFields = passports.filter { hasRequiredFields(passport: $0) }
print("There are \(passportsWithRequiredFields.count) passports with all of the required fields")

let validPassports = passports.filter { hasValidRequiredFields(passport: $0) }
print("There are \(validPassports.count) valid passports")

File: 2020/Python/04-PassportProcessing/PassportProcessing.py
#!/usr/bin/env python3

# Passport Processing
# https://adventofcode.com/2020/day/4

def read_file(name):
    file = open(name)
    return list(file.readlines())

def parse_passports(lines):
    passports = []
    passport = {}
    for line in lines:
        if not line:
            passports.append(passport)
            passport = {}
            continue
        fields = parse_fields(line)
        passport = {**passport, **fields}
    if passport:
        passports.append(passport)
    return passports

def parse_fields(line):
    fields = {}
    for field in line.split(" "):
        key, value = field.split(":")
        fields[key] = value
    return fields

def has_required_fields(passport):
    required_field_keys = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
    required_fields = [ passport[key] for key in required_field_keys if key in passport ]
    return len(required_fields) == len(required_field_keys)

def has_valid_required_fields(passport):
    required_fields = [
        ("byr", lambda value: 1920 <= int(value) <= 2002),
        ("iyr", lambda value: 2010 <= int(value) <= 2020),
        ("eyr", lambda value: 2020 <= int(value) <= 2030),
        ("hgt", lambda value: is_valid_height(value)),
        ("hcl", lambda value: is_valid_hair_color(value)),
        ("ecl", lambda value: is_valid_eye_color(value)),
        ("pid", lambda value: is_valid_passport_id(value)),
    ]
    valid_fields = [ key for key, isValid in required_fields if key in passport and isValid(passport[key]) ]
    return len(valid_fields) == len(required_fields)

def is_valid_height(value):
    height_string = value[:-2]
    if not height_string:
        return False
    height = int(value[:-2])
    return (value.endswith("in") and 59 <= height <= 76) or (value.endswith("cm") and 150 <= height <= 193)

def is_valid_hair_color(value):
    if not value.startswith("#"):
        return False
    color = value[1:]
    if len(color) != 6:
        return False
    valid = set("0123456789abcdef")
    return set(color) <= valid

def is_valid_eye_color(value):
    return value in {"amb", "blu", "brn", "gry", "grn", "hzl", "oth"}

def is_valid_passport_id(value):
    if len(value) != 9:
        return False
    valid = set("0123456789")
    return set(value) <= valid

passports = parse_passports([ line.strip() for line in read_file('04-input.txt') ])
passports_with_required_fields = [ passport for passport in passports if has_required_fields(passport) ]
print("There are {0} passports with all of the required fields".format(len(passports_with_required_fields)))

valid_passports = [ passport for passport in passports if has_valid_required_fields(passport) ]
print("There are {0} valid passports".format(len(valid_passports)))
